File: binary_search_tree/Binary_search_tree.js
class BinarySearchTree {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
  insert(value) {
    // BASE CASE
    //   if node.value is null, there is no root!
    if (!this.value) {
      this.value = value;
    } else if (value > this.value) {
      // go right. if right is empty, insert value. else recur.
      if (!this.right) {
        this.right = BinarySearchTree(value);
      } else {
      }
    } else if (value < this.value) {
      // go left. if left is empty, insert value. else recur
      if (!this.left) {
        this.left = BinarySearchTree(value);
      } else {
        this.left.insert(value);
      }
    }
  }
}

let bst = new BinarySearchTree(5);
console.log(bst);

File: binary_search_tree/binary_search_tree.py
import sys
sys.path.append('../queue_and_stack')
from dll_queue import Queue
from dll_stack import Stack


class BinarySearchTree:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

    # Insert the given value into the tree
    def insert(self, value):
        # if there is no root
        if not self.value:
            self.value = value
        # if value is smaller, go left
        elif value < self.value:
            if not self.left:
                self.left = BinarySearchTree(value)
            else:
                self.left.insert(value)
        # if value is greater, go right
        elif value >= self.value:
            if not self.right:
                self.right = BinarySearchTree(value)
            else:
                self.right.insert(value)

        # if value < root.value, go left. Set current = current.left

        # if value > root.value, go right. Set current = current.right

    # Return True if the tree contains the value
    # False if it does not
    def contains(self, target):
        current = self
        # base case
        if target == self.value:
            return True 
        elif target < self.value and self.left is None:
            return self.left.contains(target)
        elif target > self.value and self.right is None:
            return self.right.contains(target)
        return False

    # Return the maximum value found in the tree
    def get_max(self):
        if self.right is None:
            return self.value
        return self.right.get_max()

    # Call the function `cb` on the value of each node
    # You may use a recursive or iterative approach
    def for_each(self, cb):
        pass

    # DAY 2 Project -----------------------

    # Print all the values in order from low to high
    # Hint:  Use a recursive, depth first traversal
    def in_order_print(self, node):
        pass

    # Print the value of every node, starting with the given node,
    # in an iterative breadth first traversal
    def bft_print(self, node):
        pass

    # Print the value of every node, starting with the given node,
    # in an iterative depth first traversal
    def dft_print(self, node):
        pass

    # STRETCH Goals -------------------------
    # Note: Research may be required

    # Print In-order recursive DFT
    def pre_order_dft(self, node):
        pass

    # Print Post-order recursive DFT
    def post_order_dft(self, node):
        pass
